libminifi/include/utils/MinifiConcurrentQueue.h

```diff
@@ -0,0 +1,186 @@
+    std::lock_guard<std::mutex> guard(mtx_);
+    queue_.emplace_back(std::forward<Args>(args)...);
+  }
+
+ private:
+   ConcurrentQueue(ConcurrentQueue&& other, std::lock_guard<std::mutex>&)
+    : queue_( std::move(other.queue_) ) {}
+
+ protected:
+  void checkLock(std::unique_lock<std::mutex>& lck) const {
+    if (!lck.owns_lock()) {
+      throw std::logic_error("Caller of protected functions of ConcurrentQueue should own the lock!"); 
+    }
+  }
+
+  bool tryDequeueImpl(std::unique_lock<std::mutex>& lck, T& out) {
+    checkLock(lck);
+    if (queue_.empty()) {
+      return false;
+    }
+    out = std::move(queue_.front());
+    queue_.pop_front();
+    return true;
+  }
+
+  bool emptyImpl(std::unique_lock<std::mutex>& lck) const {
+    checkLock(lck);
+    return queue_.empty();
+  }
+
+  mutable std::mutex mtx_;
+ private:
+  std::deque<T> queue_;
+};
+
+
+// A ConcurrentQueue extended with a condition variable to be able to block and wait for incoming data
+template <typename T>
+class ConditionConcurrentQueue : private ConcurrentQueue<T> {
+ public:
+  explicit ConditionConcurrentQueue(bool start = true) : ConcurrentQueue<T>{}, running_{start} {}
+  
+  ConditionConcurrentQueue(const ConditionConcurrentQueue& other) = delete;
+  ConditionConcurrentQueue& operator=(const ConditionConcurrentQueue& other) = delete;
+  ConditionConcurrentQueue(ConditionConcurrentQueue&& other) = delete;
+  ConditionConcurrentQueue& operator=(ConditionConcurrentQueue&& other) = delete;
+
+  using ConcurrentQueue<T>::size;
+  using ConcurrentQueue<T>::empty;
+  using ConcurrentQueue<T>::clear;
+
+
+  template <typename... Args>
+  void enqueue(Args&&... args) {
+    ConcurrentQueue<T>::enqueue(std::forward<Args>(args)...);
+    if (running_) {
+      cv_.notify_one();
+    }
+  }
+  
+  bool dequeueWait(T& out) {
+    std::unique_lock<std::mutex> lck(this->mtx_);
+    cv_.wait(lck, [this, &lck]{ return !running_ || !this->emptyImpl(lck); });  // Only wake up if there is something to return or stopped 
+    return running_ && ConcurrentQueue<T>::tryDequeueImpl(lck, out);
+  }
+
+  template< class Rep, class Period >
+  bool dequeueWaitFor(T& out, const std::chrono::duration<Rep, Period>& time) {
+    std::unique_lock<std::mutex> lck(this->mtx_);
+    cv_.wait_for(lck, time, [this, &lck]{ return !running_ || !this->emptyImpl(lck); });  // Wake up with timeout or in case there is something to do
+    return running_ && ConcurrentQueue<T>::tryDequeueImpl(lck, out);
+  }
+
+  bool tryDequeue(T& out) {
+    std::unique_lock<std::mutex> lck(this->mtx_);
+    return running_ && ConcurrentQueue<T>::tryDequeueImpl(lck, out);
+  }
+  
+  void stop() {
+    std::lock_guard<std::mutex> guard(this->mtx_);
+    running_ = false;
+    cv_.notify_all();
+  }
+
+  void start() {
+    std::unique_lock<std::mutex> lck(this->mtx_);
+    if (!running_) {
+      running_ = true;
+      if (!this->emptyImpl(lck)) {
+        cv_.notify_all();
+      }
```

Comment:
Since we are transitioning from stopped to started in a critical section, and it's not possible to have consumers in a stopped state, this notification shouldn't be necessary.

Reply:
Removed